pytorch_memlab/line_profiler.py

```diff
@@ -58,124 +60,165 @@ def add_function(self, func):
+        refined.loc[:, acc_mask] = acc_refined
+        refined.loc[:, peak_mask] = peak_refined    
+        return refined
+
+    def records(self):
+        # Column spec: https://pytorch.org/docs/stable/cuda.html#torch.cuda.memory_stats
+        qualnames = {codehash: info['func'].__qualname__ for codehash, info in self._codes.items()}
+        records = (self._refined()
+                        .assign(qualname=lambda df: df.codehash.map(qualnames))
+                        .set_index(['qualname', 'line'])
+                        .drop(['codehash', 'num_alloc_retries', 'num_ooms'], 1))
+        records.columns = pd.MultiIndex.from_tuples([c.split('.') for c in records.columns])
+        return records
+
+    def print_stats(self, columns=['active_bytes.all.peak', 'reserved_bytes.all.peak'], func=None, stream=None):
+        if len(self._raw) == 0:
+            print('No data collected.')
+            return
 
-        if enable:
-            global_line_profiler.add_function(func)
+        records = self.records().groupby(axis=0, level=[0, 1]).max()
+        columns = [tuple(c.split('.')) for c in columns]
+        assert all(len(c) == 3 for c in columns), 'Each column name should have three dot-separated parts'
+        assert all(c in records.columns for c in columns), 'The column names should come from torch.cuda.memory_stat()\'s output'
+        records = records.loc[:, columns]
+        if func:
+            records = records.loc[[func]]
+
+        bytecols = records.columns[records.columns.get_level_values(0).str.contains('byte')]
+        maxes = records.max()
+
+        chunks = []
+        for codehash, info in self._codes.items():
+            qualname = info['func'].__qualname__
+            
+            lines, startline = inspect.getsourcelines(info['func'])
+            lines = pd.DataFrame.from_dict({'line': range(startline, startline+len(lines)), 'code': lines})
+            lines.columns = pd.MultiIndex.from_product([lines.columns, [''], ['']])
+            
+            content = pd.merge(records.loc[qualname], lines, right_on='line', left_index=True, how='right')
+            
+            style = content.style
+            for i, c in enumerate(records.columns):
+                style = style.bar([c], color=COLORS[i % len(COLORS)], width=99, vmin=0, vmax=maxes[c])
+            chunk = (style
+                        .format({c: readable_size for c in bytecols})
+                        .set_properties(subset=['code'], **{'text-align': 'left', 'white-space': 'pre', 'font-family': 'monospace'})
+                        .set_table_styles([dict(selector='th', props=[('text-align', 'left')])]) 
+                        .render())
+            chunks.append((qualname, chunk))
+
+        template = '<h3><span style="font-family: monospace">{q}</span></h3><div>{c}</div>'
+        html = '\n'.join(template.format(q=q, c=c) for q, c in chunks)
+        if stream is None:
+            display(HTML(html))
```

Comment:
Will this support printing to the command line interface?

Reply:
No. Let's discuss back in the main thread what to do about the command line.